### PR TITLE
feat: add new API endpoint to fetch network ID

### PR DESCRIPTION
**Pull request by mvadari**
## High Level Overview of Change

This PR creates a new API endpoint, `get_network`, that fetches the network ID of the network that a node is connected to. If it doesn't exist, it creates one, and immediately crawls the network.

The API endpoint takes 300-40ms to load for a rippled node that exists, and 6ish seconds for an endpoint that is not a rippled node (since it waits for the call to `/crawl` to error).

### Context of Change

This is necessary for sidechains - if the Explorer wants to visit a new sidechain that it hasn't been to before, it needs to add it to the VHS. And this must happen programatically.

### Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Test Plan

CI passes. Tested several scenarios by hand.

### CODE REVIEW

File: src/api/routes/v1/get-network.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+ */
+async function addNode(url: string, unl: string | null): Promise<string> {
+  const currentNetworks = await query('networks')
+    .select('network')
+    .orderBy('network')
+  const maxNetwork =
+    currentNetworks[currentNetworks.length - networks.length - 1]?.network ?? 0
+  const newNetwork = (Number(maxNetwork) + 1).toString()
+  const network: Network = {
+    network: newNetwork,
```
Comment (JST5000) [inline src/api/routes/v1/get-network.ts:134]: Nit: I got a bit confused reading this because I couldn't tell if `Network.network` was supposed to be an integer or a string. Maybe a better name for that field would be `id`?

File: src/api/routes/v1/get-network.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+}
+
+/**
+ * Add entry to new network.
+ *
+ * @param req - Express request.
+ * @param res - Express response.
+ * @returns The Express response.
+ */
+export default async function getNetwork(
```
Comment (JST5000) [inline src/api/routes/v1/get-network.ts:157]: I wouldn't expect `getNetwork` to modify the network - Maybe `putNetwork` would be a better name?

Comment (JST5000) [reply src/api/routes/v1/get-network.ts:157]: (Or maybe the description for this function should make it sound more like it's caching the result instead of adding a new network)

**Review by JST5000 — APPROVED** [review]
Logic looks good to me - Just a couple naming thoughts :)

File: src/api/routes/v1/get-network.ts [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        created: false,
+      })
+    }
+    // add node to networks list
+    const newNetwork = await addNode(entryUrl, node_unl)
+
+    return res.send({
+      result: 'success',
+      network: newNetwork,
+      created: true,
```
Comment (natenichols) [inline src/api/routes/v1/get-network.ts:197]: I don't necessarily think we have to tell the user if an entry has been created.

**Review by natenichols — APPROVED** [review]

File: src/api/routes/v1/get-network.ts (hunk 1 above)
Comment (mvadari) [reply src/api/routes/v1/get-network.ts:140]: It's referred to as `network` in a bunch of the existing tables, which is why I kept that as `id`. But happy to rename it in this one if you'd prefer.

Comment (mvadari) [reply src/api/routes/v1/get-network.ts:140]: It's actually a string (it's `main`/`test/`dev/`nft-dev` for the existing networks, and I didn't want to mess with the existing tables too much bc migrations by hand are hard and scary). But I thought the best way to add a new network without any details about it would be to just assign it a number.

Comment (mvadari) [reply src/api/routes/v1/get-network.ts:140]: `currentNetworks.length - networks.length - 1` handles that - non-number strings will be after the number strings when sorted alphabetically

Comment (mvadari) [reply src/api/routes/v1/get-network.ts:140]: Hm, looks like I'll have to rethink the sorting anyways - I forgot that with strings, '2' comes after '11'

File: src/api/routes/v1/get-network.ts (hunk 2 above)
Comment (JST5000) [reply src/api/routes/v1/get-network.ts:163]: That's fair, only name I could think of is `fetchNetwork`, but that's not much better - just the description update is clear enough for me :)

File: src/api/routes/v1/get-network.ts (hunk 1 above)
Comment (JST5000) [reply src/api/routes/v1/get-network.ts:140]: I think it'd be worth adding a comment explaining that going to `currentNetworks.length - networks.length - 1` is to go to the end of the named networks (`networks.length`) and into the numbered networks.

**Review by ckniffen — APPROVED** [review]

File: src/api/routes/v1/get-network.ts (hunk 2 above)
Comment (mvadari) [reply src/api/routes/v1/get-network.ts:163]: What about something like `getNetworkOrAdd`?

IMO the method name isn't that important, the API endpoint matters more

File: src/crawler/index.ts [hunk 4]
```diff
@@ -49,4 +49,6 @@ async function main(): Promise<void> {
   }, LOCATE_INTERVAL)
 }
 
-main().catch((err: Error) => log.error(`${err.message}`))
+if (typeof require !== 'undefined' && require.main === module) {
```
Comment (ckniffen) [inline src/crawler/index.ts:52]: Why is this required now?

**Review by ckniffen — APPROVED** [review]

File: src/api/routes/v1/get-network.ts (hunk 2 above)
Comment (JST5000) [reply src/api/routes/v1/get-network.ts:194]: I think `getNetworkOrAdd` conveys the functionality more clearly

File: src/api/routes/v1/get-network.ts [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+import { Crawl } from '../../../shared/types'
+import logger from '../../../shared/utils/logger'
+
+const log = logger({ name: 'api-get-network' })
+
+const CRAWL_PORTS = [51235, 2459, 30001]
+
+let maxNetwork: number
+
+async function getMaxNetwork(): Promise<void> {
```
Comment (JST5000) [inline src/api/routes/v1/get-network.ts:16]: `getMaxNetwork` implies that it will return maxNetwork, I think this name better reflects it's behavior.
```suggestion
async function updateMaxNetwork(): Promise<void> {
```

File: src/api/routes/v1/get-network.ts [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+/**
+ * Adds the node (and its corresponding network) to the networks table.
+ *
+ * @param url - The URL endpoint of the node.
+ * @param unl - The UNL of the node.
+ * @returns The ID of the new network.
+ */
+async function addNode(url: string, unl: string | null): Promise<string> {
+  const newNetwork = (Number(maxNetwork) + 1).toString()
```
Comment (JST5000) [inline src/api/routes/v1/get-network.ts:167]: Why do you have to put `Number(maxNetwork)` instead of `maxNetwork` on its own?

File: src/api/routes/v1/get-network.ts [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    const { public_key } = crawl.this_node
+    const publicKeyNetwork = await getNetworkFromPublicKey(public_key)
+    if (publicKeyNetwork != null) {
+      return res.send({
+        result: 'success',
+        network: publicKeyNetwork,
+      })
+    }
+    // add node to networks list
+    const newNetwork = await addNode(entryUrl, node_unl)
```
Comment (natenichols) [inline src/api/routes/v1/get-network.ts:227]: Can we add a check that the crawl was valid (i.e. produced > `n`) nodes. Otherwise I think I could just spam random entries that will be saved in `addNode`. 

Also, I think `node_unl` can be undefined here (if node_unl == null). Should that be a possibility, or can we save an IP w/o a node. 

We should ensure that we got a valid crawl, a valid UNL, and a valid public_key before we save a node, else anyone can write arbitrary information to the database.

File: src/api/routes/v1/get-network.ts [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+      if (unlNetwork != null) {
+        return res.send({
+          result: 'success',
+          network: unlNetwork,
+        })
+      }
+    }
+
+    // check if node public key is already recorded
+    const { public_key } = crawl.this_node
```
Comment (natenichols) [inline src/api/routes/v1/get-network.ts:218]: is `this_node` guaranteed to exist? If not this will error:
```js
const foo = {}
const { does_not_exist } = foo.not_a_field
```
Because that the same as
```js
const does_not_exists = foo.not_a_field.does_not_exist
``` 

and JS doesn't let you call `.does_not_exist` on undefined.

File: src/api/routes/v1/get-network.ts (hunk 7 above)
Comment (natenichols) [reply src/api/routes/v1/get-network.ts:227]: I would also err on the side of making it more difficult to add your chain to the VHS database, over making it too easy.

File: src/api/routes/v1/get-network.ts [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+    port: 51235,
+    unls: unl ? [unl] : [],
+  }
+  await query('networks').insert({
+    ...network,
+    unls: network.unls.join(','),
+  })
+
+  const crawler = new Crawler()
+  void crawler.crawl(network)
```
Comment (natenichols) [inline src/api/routes/v1/get-network.ts:182]: So we kick off a crawl for that network to find new nodes for the newly added Node?

**Review by natenichols — CHANGES_REQUESTED** [review]
This looks great! I have a couple questions about `getNetwork(` specifically, since that seems the core functionality that gets added.

Happy to approve once questions are answered.

File: src/crawler/index.ts (hunk 4 above)
Comment (mvadari) [reply src/crawler/index.ts:52]: It was running when I imported from this file. This prevents that from happening (it's the JS equivalent of `if __name__ == '__main__'` in python).

File: src/api/routes/v1/get-network.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+      })
+    }
+    // add node to networks list
+    const newNetwork = await addNode(entryUrl, node_unl)
+
+    return res.send({
+      result: 'success',
+      network: newNetwork,
+    })
+  } catch (err) {
```
[the remark this answers is not in the document]
Comment (mvadari) [reply src/api/routes/v1/get-network.ts:233]: The app doesn't crash, but it does return an error if it's not caught.

File: src/api/routes/v1/get-network.ts (hunk 9 above)
Comment (mvadari) [reply src/api/routes/v1/get-network.ts:182]: Yes. Otherwise, you'll have to wait until the next set of crawls before you have the nodes from the network.

File: src/api/routes/v1/get-network.ts (hunk 7 above)
Comment (mvadari) [reply src/api/routes/v1/get-network.ts:227]: I don't quite understand how you could spam random entries - the `getNetworkFromPublicKey` call checks whether the node already exists in the known networks.

File: src/api/routes/v1/get-network.ts (hunk 8 above)
Comment (mvadari) [reply src/api/routes/v1/get-network.ts:218]: If it's a valid crawl, it should. If not, I believe it'll error in the [`crawlNode` function in the crawler](https://github.com/ripple/validator-history-service/blob/95ba58da44d82308e34c169e79cdad109528ab50/src/crawler/network.ts#L41-L95), where the crawl is fetched.

**Review by natenichols — APPROVED** [review]
